### ISSUE TITLE
Holistic per-backend attribute system

### ISSUE THREAD

**Issue by Manishearth** [unicode-org/icu4x#2453]
I'd like to have a single system for passing attributes to backends.

Right now we have a couple proposed:

 - `#[diplomat::namespace(cpp, icu4x)]`: https://github.com/rust-diplomat/diplomat/issues/103
 - `#[diplomat::rename(cpp, foo_bar)]` https://github.com/unicode-org/icu4x/issues/2451
 - `#[diplomat::disable(cpp)]`: https://github.com/unicode-org/icu4x/issues/2450

We may also want to have a consistent way for backends to do custom stuff, e.g. `#[diplomat::backend_specific(cpp, capture = by_ref)]`

Ideally this all goes through the same underlying infrastructure, such that backends get a filtered list of attributes. Ideally this lets us be more robust in what we support as the first argument here; for example, I'm imagining:

 - `not()`: #[diplomat::namespace(not(js, java), icu4x)]` (multiple arguments is an alias for `not(any())`)
 - `any()`/`all()`: #[diplomat::namespace(any(c, cpp), icu4x)]`
 - `overloadable`: `#[diplomat::rename(overloadable, foo_bar)]`
    - This would only pass down to backends that claim to support overloading. Not every backend will support the same set of overloads, but this is a convenient default to apply to overloads and when an overload is not supported by a backend the error will give the user an opportunity to figure out how to handle overloading and write the right API for it
 - `constructors`: `#[diplomat::rename(constructors, constructor)]`: backends that support the `constructor` rename to create a constructor
 - Potentially allow people to pass in their own CFG keywords, I can imagine ICU4X using keywords to talk about groups of backends instead of peppering the codebase with `any(cpp, js)` or whatever. With this, basically, `c`, `cpp`, `js`, `overloadable`, etc are just keywords that the backend autosets for you, there's nothing actually specia about backends.

Basically a full cfg language with some magic keywords

**Comment by Manishearth** [comment]
Wrong repo, https://github.com/rust-diplomat/diplomat/issues/235
